### PR TITLE
Adding system agent support

### PR DESCRIPTION
**Pull request by phillipsj**
This is the basic work to get system agent running in Wins.

### CODE REVIEW

File: cmd/server/app/run.go [hunk 1]
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"time"
 
+	"github.com/rancher/wins/pkg/sa"
+
```
Comment (luthermonson) [inline cmd/server/app/run.go:8]: goimports

**Review by rosskirkpat — APPROVED** [review]
LGTM

**Review by brandond — APPROVED** [review]

**Review by dweomer — APPROVED** [review]

File: cmd/server/app/run_service.go [hunk 2]
```diff
@@ -2,6 +2,7 @@ package app
 
 import (
 	"context"
+	"github.com/rancher/wins/pkg/sa"
```
Comment (nickgerace) [inline cmd/server/app/run_service.go:5]: nit: We should move this dep

File: cmd/server/app/run_service.go [hunk 3]
```diff
@@ -241,6 +244,11 @@ func (h *serviceHandler) Execute(_ []string, r <-chan svc.ChangeRequest, s chan<
 		h.errC <- h.srv.Serve(ctx)
 	}()
 
+    // start system agent
+	go func() {
+		h.errC <- h.agent.Run(ctx)
+	}()
```
Comment (nickgerace) [inline cmd/server/app/run_service.go:250]: nit: `fmt`

File: pkg/sa/agent.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+type Agent struct {
+	cfg *config.AgentConfig
+}
+
+func (a *Agent) Run(ctx context.Context) error {
+
+	if a.cfg == nil {
+		return nil
```
Comment (nickgerace) [inline pkg/sa/agent.go:22]: Should we error out with a nil config instead? What's the real world scenario when this is ran with a nil config?

File: pkg/sa/agent.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+		var connInfo config.ConnectionInfo
+
+		if err := config.Parse(a.cfg.ConnectionInfoFile, &connInfo); err != nil {
+			logrus.Fatalf("Unable to parse connection info file %v", err)
+		}
+
+		k8splan.Watch(ctx, *applier, connInfo)
+	}
+
+	if a.cfg.LocalEnabled {
```
Comment (nickgerace) [inline pkg/sa/agent.go:48]: nit: `else if` might be better here for readability

File: pkg/sa/agent.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	if a.cfg.RemoteEnabled {
+		logrus.Infof("Starting remote watch of plans")
+
+		var connInfo config.ConnectionInfo
+
+		if err := config.Parse(a.cfg.ConnectionInfoFile, &connInfo); err != nil {
+			logrus.Fatalf("Unable to parse connection info file %v", err)
+		}
+
+		k8splan.Watch(ctx, *applier, connInfo)
```
Comment (nickgerace) [inline pkg/sa/agent.go:45]: Does this watcher handle its own errors?

File: pkg/sa/agent.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	imageUtil := image.NewUtility(a.cfg.ImagesDir, a.cfg.ImageCredentialProviderConfig, a.cfg.ImageCredentialProviderBinDir, a.cfg.AgentRegistriesFile)
+	applier := applyinator.NewApplyinator(a.cfg.WorkDir, a.cfg.PreserveWorkDir, a.cfg.AppliedPlanDir, imageUtil)
+
+	if a.cfg.RemoteEnabled {
+		logrus.Infof("Starting remote watch of plans")
+
+		var connInfo config.ConnectionInfo
+
+		if err := config.Parse(a.cfg.ConnectionInfoFile, &connInfo); err != nil {
+			logrus.Fatalf("Unable to parse connection info file %v", err)
```
Comment (nickgerace) [inline pkg/sa/agent.go:42]: Why not return this? We need to exit?

File: pkg/sa/agent.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	}
+
+	return nil
+}
+
+func New(cfg *config.AgentConfig) (*Agent, error) {
+	return &Agent{
+		cfg: cfg,
+	}, nil
+}
```
Comment (nickgerace) [inline pkg/sa/agent.go:60]: This function doesn't seem necessary to me. We don't need to future proof an error statement IMO.

**Review by nickgerace — CHANGES_REQUESTED** [review]

File: pkg/sa/agent.go (hunk 4 above)
Comment (phillipsj) [reply pkg/sa/agent.go:22]: If the agent config isn't set, aka nil, then it's a no op. There is a decision record for t design.

File: pkg/sa/agent.go (hunk 5 above)
Comment (phillipsj) [reply pkg/sa/agent.go:48]: It's how it is used upstream in system agent.

File: pkg/sa/agent.go (hunk 6 above)
Comment (phillipsj) [reply pkg/sa/agent.go:45]: This is how it's used upstream and it does handle errors.

File: docs/adrs/0002-add-system-agent-support-to-wins.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+Accepted
+
+## Context
+
+The [System Agent]() is used to install and upgrade RKE2 using plans. Support is needed to provide this ability for Windows. 
+
+## Decision
+
+System Agent functionality needs some capabilities of Wins to achieve its goal. This would require that both the system agent service and wins service to be running on a system. Alternatively, we could consume the system agent as a package in the wins service reducing the need to having both installed in a system.
```
Comment (aiyengar2) [inline docs/adrs/0002-add-system-agent-support-to-wins.md:15]: nit: Seems like this should be part of the Context, not the decision. The decision is just the last sentence right?

File: pkg/sa/agent.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+func (a *Agent) Run(ctx context.Context) error {
+
+	if a.cfg == nil {
+		return nil
+	}
+
+	logrus.Infof("Rancher System Agent version %s is starting", version.FriendlyVersion())
+
+	if !a.cfg.LocalEnabled && !a.cfg.RemoteEnabled {
+		logrus.Fatalf("Local and remote were both not enabled. Exiting, as one must be enabled.")
```
Comment (aiyengar2) [inline pkg/sa/agent.go:28]: Should the behavior here also be to return `nil`, based on https://github.com/rancher/wins/pull/56/files#r721730809? If so, should we move this before the `logrus` line?

File: pkg/sa/agent.go [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+	}
+
+	logrus.Infof("Using directory %s for work", a.cfg.WorkDir)
```
Comment (aiyengar2) [inline pkg/sa/agent.go:31]: nit: 
```suggestion
	logrus.Infof("Setting %s as the working directory", a.cfg.WorkDir)
```

File: pkg/sa/agent.go (hunk 7 above)
Comment (aiyengar2) [reply pkg/sa/agent.go:42]: should we add a colon before the `%v`?

File: pkg/sa/agent.go (hunk 8 above)
Comment (aiyengar2) [reply pkg/sa/agent.go:60]: Agreed. I don't think this is necessary.

**Review by aiyengar2 — CHANGES_REQUESTED** [review]

File: pkg/sa/agent.go (hunk 5 above)
Comment (nickgerace) [reply pkg/sa/agent.go:48]: My mistake, I did not realize you could use both at once.

File: pkg/sa/agent.go (hunk 7 above)
Comment (nickgerace) [reply pkg/sa/agent.go:42]: ^ that's a `nit` if I've ever seen one 
image

**Review by nickgerace — APPROVED** [review]

**Review by aiyengar2 — APPROVED** [review]
LGTM

File: cmd/server/app/run.go [hunk 12]
```diff
@@ -104,7 +105,12 @@ func _runAction(cliCtx *cli.Context) error {
 	if err != nil {
 		return errors.Wrapf(err, "failed to create server")
 	}
-	err = runService(ctx, server)
+	agent := systemagent.New(cfg.SystemAgent)
+	if err != nil {
```
Comment (briandowns) [inline cmd/server/app/run.go:109]: What error is being checked here?

File: cmd/server/app/run_service.go [hunk 13]
```diff
@@ -241,6 +244,11 @@ func (h *serviceHandler) Execute(_ []string, r <-chan svc.ChangeRequest, s chan<
 		h.errC <- h.srv.Serve(ctx)
 	}()
 
+	// start system agent
+	go func() {
+		h.errC <- h.agent.Run(ctx)
```
Comment (briandowns) [inline cmd/server/app/run_service.go:249]: Should we be mixing the use of this channel wtih agent and server?

Comment (phillipsj) [reply cmd/server/app/run_service.go:249]: That channel is for the windows service, so both "sub processes" should be reporting into that channel.

**Review by briandowns — APPROVED** [review]

**Pull request merged**